File: packages/guards/src/chunk-boundary-leak.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface ChunkBoundaryLeakOptions {
  action: 'block' | 'warn';
  customSeparators?: string[];
}

const DEFAULT_SEPARATORS = [
  '---', '===', '###END###',
  'Document:', 'Chunk:', 'Source:',
  '<|endoftext|>', '[SEP]', '<<CONTEXT>>',
  'metadata:', 'chunk_id:', 'embedding:',
  'score:', 'relevance_score:',
  '\n---\n', 'Retrieved from:', 'passage_id:',
];

export function chunkBoundaryLeak(options: ChunkBoundaryLeakOptions): Guard {
  const separators = options.customSeparators
    ? [...DEFAULT_SEPARATORS, ...options.customSeparators]
    : DEFAULT_SEPARATORS;

  return {
    name: 'chunk-boundary-leak',
    version: '0.1.0',
    description: 'Detects RAG chunk metadata or separators leaking into output',
    category: 'security',
    supportedStages: ['output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const lower = text.toLowerCase();
      const found: string[] = [];

      for (const sep of separators) {
        if (lower.includes(sep.toLowerCase())) {
          found.push(sep);
        }
      }

      const triggered = found.length > 0;

      return {
        guardName: 'chunk-boundary-leak',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        latencyMs: Math.round(performance.now() - start),
        details: triggered ? { leakedSeparators: found } : undefined,
      };
    },
  };
}

File: python/open_guardrail/guards/chunk_boundary_leak.py
"""Detect RAG chunk metadata or separators leaking into output."""
from __future__ import annotations

import time
from typing import List, Optional

from open_guardrail.core import GuardResult

_DEFAULT_SEPARATORS = [
    "---", "===", "###END###",
    "Document:", "Chunk:", "Source:",
    "<|endoftext|>", "[SEP]", "<<CONTEXT>>",
    "metadata:", "chunk_id:", "embedding:",
    "score:", "relevance_score:",
    "\n---\n", "Retrieved from:", "passage_id:",
]


class _ChunkBoundaryLeak:
    def __init__(
        self,
        *,
        action: str = "warn",
        custom_separators: Optional[List[str]] = None,
    ) -> None:
        self.name = "chunk-boundary-leak"
        self.action = action
        self._separators = (
            _DEFAULT_SEPARATORS + custom_separators
            if custom_separators
            else _DEFAULT_SEPARATORS
        )

    def check(
        self, text: str, stage: str = "output"
    ) -> GuardResult:
        start = time.perf_counter()
        lower = text.lower()
        found: list[str] = []

        for sep in self._separators:
            if sep.lower() in lower:
                found.append(sep)

        triggered = len(found) > 0
        elapsed = (time.perf_counter() - start) * 1000

        return GuardResult(
            guard_name="chunk-boundary-leak",
            passed=not triggered,
            action=self.action if triggered else "allow",
            message=(
                "Chunk boundary separators leaked"
                if triggered
                else None
            ),
            latency_ms=round(elapsed, 2),
            details=(
                {"leaked_separators": found}
                if triggered
                else None
            ),
        )


def chunk_boundary_leak(
    *,
    action: str = "warn",
    custom_separators: Optional[List[str]] = None,
) -> _ChunkBoundaryLeak:
    return _ChunkBoundaryLeak(
        action=action,
        custom_separators=custom_separators,
    )
